File: frontend/src/types/insights.ts
// frontend/src/types/insights.ts
//
// Mirrors backend/app/schemas/insights.py. Ground truth for field names is
// that file's docstring, which itself defers to mock-src/fixtures.py where
// the two disagree — see the NOTE comments there.
//
// House rules carried over from the backend:
// - Money is ALWAYS a fixed-2-decimal STRING at the JSON boundary
//   (e.g. "1234.56"), never a number. Parse with `parseMoney` from
//   `@/lib/insights-utils` only where a computation actually needs a number.
// - Ratios/percentages (delta_pct, progress, shares, breakeven_discount) are
//   plain floats — not money, no string encoding.
// - The tab never computes date arithmetic client-side: days_since_review,
//   coverage pct, etc. all arrive pre-computed from the server.

/** Fixed-2-decimal string, e.g. "1234.56" or "-42.00". Never a number. */
export type Money = string

// ---------------------------------------------------------------------------
// Error envelope
// ---------------------------------------------------------------------------

export type InsightsErrorCode =
  | 'INSUFFICIENT_HISTORY'
  | 'NOT_CLASSIFIED'
  | 'NO_TRUSTED_WINDOW'
  | 'NO_YIELD_BASIS'
  | 'COMPUTATION_FAILED'

export interface InsightsError {
  code: InsightsErrorCode
  message: string
  retryable: boolean
}

export interface InsightsWindow {
  // Serialized as `from` on the wire (populate_by_name alias in the backend
  // model) — `from` is a reserved word in some JS/TS contexts (e.g. as an
  // identifier in certain grammars, and it shadows Array.from-style usage
  // awkwardly), so the field is kept as `from` here since it's valid as an
  // object property key/string literal in TypeScript; access it via
  // `window.from` or destructure with `{ from: windowFrom }`.
  from: string
  to: string
  trusted_from: string | null
  months_trusted: number
}

export type InsightsReference = 'budget' | 'historical'

export interface InsightsEnvelope<T> {
  data: T | null
  error: InsightsError | null
  window: InsightsWindow | null
  reference: InsightsReference | null
  currency: string
  generated_at: string
}

// ---------------------------------------------------------------------------
// 1 - Alerts
// ---------------------------------------------------------------------------

export interface AlertLink {
  view: string
  params: Record<string, string>
}

export interface AlertRow {
  id: string
  severity: 'crit' | 'warn' | 'info'
  kind: 'overspend' | 'hygiene' | 'cash_gap' | 'goal_off_track'
  title: string
  detail: string
  amount: Money | null
  link: AlertLink | null
}

export type AlertsData = AlertRow[]

// ---------------------------------------------------------------------------
// 2 - Vitals
// ---------------------------------------------------------------------------

export interface VitalReference {
  value: Money
  source: 'budget' | 'historical' | 'benchmark'
  method: string | null
  label: string
}

export interface VitalSeriesPoint {
  date: string
  value: Money
  trusted: boolean
}

export interface VitalCard {
  key: 'runway' | 'savings_rate' | 'net_worth' | 'credit_utilization'
  label: string
  value: Money | null
  unit: 'BRL' | 'months' | 'percent'
  reference: VitalReference | null
  status: 'good' | 'warn' | 'crit' | 'unknown'
  available: boolean
  blocked_reason: string | null
  series: VitalSeriesPoint[] | null
}

export type VitalsData = VitalCard[]

// ---------------------------------------------------------------------------
// 3 - Flow (Sankey)
// ---------------------------------------------------------------------------

export interface FlowNode {
  id: string
  label: string
  depth: 0 | 1 | 2
  kind: 'income' | 'group' | 'category' | 'saved'
  color: string
  value: Money
}

export interface FlowLink {
  source: string
  target: string
  value: Money
}

export interface FlowData {
  nodes: FlowNode[]
  links: FlowLink[]
  collapse_threshold: Money
  // Optional: the empty-month fixture omits these entirely even though the
  // design brief says they should "always" be present. Backend schema makes
  // them optional to tolerate that; do the same here.
  income_total: Money | null
  deficit: Money | null
}

// ---------------------------------------------------------------------------
// 4 - Nature (fixed / variable / discretionary)
// ---------------------------------------------------------------------------

export interface NatureShares {
  fixed: number
  variable: number
  discretionary: number
  unclassified: number
}

export interface NatureValues {
  fixed: Money
  variable: Money
  discretionary: Money
  unclassified: Money
}

export interface NatureMonth {
  month: string
  trusted: boolean
  income: Money | null
  values: NatureValues
  shares: NatureShares | null
}

export interface NatureReference {
  fixed: Money
  variable: Money
  discretionary: Money
  unclassified: Money
}

export interface NatureData {
  series: NatureMonth[]
  // NOTE: fixtures.py never populates this even though contratos.html's
  // "Forma do dado" lists it at the top level. Defensive: treat as possibly
  // absent/null.
  reference: NatureReference | null
}

// ---------------------------------------------------------------------------
// 5 - Projection
// ---------------------------------------------------------------------------

export interface ProjectionComponents {
  income_expected: Money
  recurring: Money
  installments: Money
  variable_estimate: Money
}

export interface ProjectionPoint {
  month: string
  kind: 'actual' | 'projected'
  balance: Money
  low: Money | null
  high: Money | null
  committed: Money
  components: ProjectionComponents
}

export interface ProjectionAssumption {
  label: string
  value: string
  source: string
}

export interface ProjectionData {
  points: ProjectionPoint[]
  assumptions: ProjectionAssumption[]
}

// ---------------------------------------------------------------------------
// 6 - Categories table
// ---------------------------------------------------------------------------

export interface CategoryRow {
  category_id: string | null
  label: string
  color: string
  group_label: string
  nature: 'fixed' | 'variable' | 'discretionary' | null
  amount: Money
  reference: Money | null
  reference_method: string | null
  delta: Money | null
  delta_pct: number | null
  tx_count: number
  status: 'over' | 'under' | 'ok' | 'no_ref'
}

export interface CategoriesData {
  rows: CategoryRow[]
  available: boolean
  blocked_reason: string | null
}

// ---------------------------------------------------------------------------
// 7 - Hygiene
// ---------------------------------------------------------------------------

export interface HygieneCoverageStat {
  done: number
  total: number
  // Server-computed (a `computed_field` on the backend model) — never derive
  // this client-side, per the "screen never computes" house rule.
  pct: number
}

export interface HygieneCoverage {
  categorized: HygieneCoverageStat
  nature_set: HygieneCoverageStat
  transfers_paired: HygieneCoverageStat
}

export interface HygieneOffender {
  kind: 'month' | 'account' | 'payee'
  label: string
  missing: number
  amount: Money
}

export interface HygieneData {
  coverage: HygieneCoverage
  last_review_at: string | null
  // NOTE: fixtures.py never populates this field even though the design
  // brief and contratos.html both call for it. Treat as possibly null/absent
  // and never derive it client-side from last_review_at (that would violate
  // the "client must never divide or subtract dates" rule the field exists
  // to satisfy in the first place).
  days_since_review: number | null
  worst_offenders: HygieneOffender[]
}

// ---------------------------------------------------------------------------
// 8 - Goals
// ---------------------------------------------------------------------------

export interface GoalRow {
  goal_id: string
  label: string
  current: Money
  target: Money
  progress: number
  target_date: string | null
  observed_contribution: Money | null
  estimated_completion: string | null
  required_contribution: Money | null
  status: 'on_track' | 'behind' | 'stalled' | 'unknown'
  blocked_reason: string | null
}

export type GoalsData = GoalRow[]

// ---------------------------------------------------------------------------
// 9 - Breakeven table + purchase decision calculator
// ---------------------------------------------------------------------------

export interface YieldRange {
  p25: string
  p75: string
}

export interface YieldWindowDays {
  min: number
  max: number
}

export interface YieldBasis {
  monthly_gross: string
  method: 'observed'
  sample_size: number
  window_days: YieldWindowDays
  range: YieldRange
  as_of: string
}

export interface BreakevenRow {
  n: number
  ir_rate: number
  net_monthly: string
  breakeven_discount: number
  gain_per_1000: Money
}

export interface BreakevenTableData {
  yield_basis: YieldBasis
  rows: BreakevenRow[]
}

export interface PurchaseDecisionRequest {
  price: Money
  cash_discount_pct: number
  installments: number
  yield_override: number | null
}

export interface PurchaseDecisionVerdict {
  choice: 'installments' | 'cash'
  net_gain: Money
  breakeven_discount: number
  headline: string
}

export interface PurchaseDecisionScheduleRow {
  month: number
  payment: Money
  balance_invested: Money
  yield_net: Money
  cumulative_gain: Money
}

export interface FlexibilityMonthlyCommitmentDelta {
  month: string
  // NOTE: unconfirmed shape — contratos.html only sketches `{month, ...}`
  // and no fixture exercises this endpoint (it's a live POST calculator).
  delta: Money
}

export interface FlexibilityImpact {
  utilization_before: number
  utilization_after: number
  monthly_commitment_delta: FlexibilityMonthlyCommitmentDelta[]
  cash_gap_month: string | null
}

export interface PurchaseDecisionData {
  verdict: PurchaseDecisionVerdict
  schedule: PurchaseDecisionScheduleRow[]
  flexibility_impact: FlexibilityImpact
}

File: backend/app/schemas/insights.py
"""Schemas for the Insights tab (GET /api/insights/*).

Ground truth for field names and nesting is the executable fixtures mock:
mock-src/fixtures.py (see finance/insights-design/). Where this file and the
original design brief disagreed, the fixtures won — discrepancies are noted
inline as comments prefixed with `NOTE:`.

House rules baked into these schemas (see contratos.html):
- Money is always a fixed-2-decimal string in JSON, Decimal internally.
- Ratios/percentages (delta_pct, progress, shares, breakeven_discount) are
  floats — they are not money and don't share its precision problem.
  Frequencies (0.20 for 20%) are represented as floats in [0, 1] unless the
  fixtures show otherwise (they do not always agree, see NOTE below).
- The tab never computes: days_since_review, coverage pct, etc. all arrive
  pre-computed from the server.
- Every block-level response optionally carries `window`/`reference`; the
  calculator (breakeven-table, purchase-decision) and a couple of blocks
  without a reference axis (flow, projection, hygiene) omit them.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Generic, Literal, TypeVar

from pydantic import BaseModel, Field, PlainSerializer, computed_field

T = TypeVar("T")

# ---------------------------------------------------------------------------
# Money: Decimal internally, fixed 2-decimal string at the JSON boundary.
# ---------------------------------------------------------------------------

Money = Annotated[Decimal, PlainSerializer(lambda v: f"{v:.2f}", return_type=str)]


# ---------------------------------------------------------------------------
# Error envelope
# ---------------------------------------------------------------------------

# NOTE: contratos.html's decided error contract only lists these codes:
# it doesn't enumerate a closed set of `code` values, it just says "a code
# the screen decides by". The brief asked for a fixed set of five codes;
# fixtures.py never emits an error envelope at all (every scenario returns
# `data` instead), so there is no ground truth for the exact code strings.
# Following the brief's five codes as the closed set since it's the only
# guidance available; COMPUTATION_FAILED is the only retryable one.
InsightsErrorCode = Literal[
    "INSUFFICIENT_HISTORY",
    "NOT_CLASSIFIED",
    "NO_TRUSTED_WINDOW",
    "NO_YIELD_BASIS",
    "COMPUTATION_FAILED",
]


class InsightsError(BaseModel):
    code: InsightsErrorCode
    message: str
    retryable: bool = False


class InsightsWindow(BaseModel):
    model_config = {"populate_by_name": True}

    from_: date = Field(alias="from")
    to: date
    trusted_from: date | None
    months_trusted: int


class InsightsEnvelope(BaseModel, Generic[T]):
    data: T | None
    error: InsightsError | None = None
    window: InsightsWindow | None = None
    reference: Literal["budget", "historical"] | None = None
    currency: str
    generated_at: datetime


# ---------------------------------------------------------------------------
# 1 - Alerts
# ---------------------------------------------------------------------------


class AlertLink(BaseModel):
    view: str
    params: dict[str, str]


class AlertRow(BaseModel):
    id: str
    severity: Literal["crit", "warn", "info"]
    kind: Literal["overspend", "hygiene", "cash_gap", "goal_off_track"]
    title: str
    detail: str
    amount: Money | None
    link: AlertLink | None


# `alerts` in fixtures.py is a bare list, not wrapped in an extra object.
AlertsData = list[AlertRow]


# ---------------------------------------------------------------------------
# 2 - Vitals
# ---------------------------------------------------------------------------


class VitalReference(BaseModel):
    value: Money
    source: Literal["budget", "historical", "benchmark"]
    method: str | None
    label: str


class VitalSeriesPoint(BaseModel):
    date: str
    value: Money
    trusted: bool


class VitalCard(BaseModel):
    key: Literal["runway", "savings_rate", "net_worth", "credit_utilization"]
    label: str
    value: Money | None
    unit: Literal["BRL", "months", "percent"]
    reference: VitalReference | None
    status: Literal["good", "warn", "crit", "unknown"]
    available: bool
    blocked_reason: str | None
    series: list[VitalSeriesPoint] | None


# `vitals` in fixtures.py is a bare list of the four cards.
VitalsData = list[VitalCard]


# ---------------------------------------------------------------------------
# 3 - Flow (Sankey)
# ---------------------------------------------------------------------------


class FlowNode(BaseModel):
    id: str
    label: str
    depth: Literal[0, 1, 2]
    kind: Literal["income", "group", "category", "saved"]
    color: str
    value: Money


class FlowLink(BaseModel):
    source: str
    target: str
    value: Money


class FlowData(BaseModel):
    nodes: list[FlowNode]
    links: list[FlowLink]
    collapse_threshold: Money
    # NOTE: fixtures.py only sets income_total/deficit in the non-empty
    # branch; the "vazio" (empty) scenario returns
    # dict(nodes=[], links=[], collapse_threshold="0.00") with no
    # income_total/deficit keys at all. contratos.html says deficit must
    # "sempre vir" (always be present), so we make them optional here
    # rather than required, to tolerate the empty-month case the fixtures
    # actually produce.
    income_total: Money | None = None
    deficit: Money | None = None


# ---------------------------------------------------------------------------
# 4 - Nature (fixed / variable / discretionary)
# ---------------------------------------------------------------------------


class NatureShares(BaseModel):
    fixed: float
    variable: float
    discretionary: float
    unclassified: float


class NatureValues(BaseModel):
    fixed: Money
    variable: Money
    discretionary: Money
    unclassified: Money


class NatureMonth(BaseModel):
    month: str
    trusted: bool
    income: Money | None
    values: NatureValues
    shares: NatureShares | None


class NatureReference(BaseModel):
    fixed: Money
    variable: Money
    discretionary: Money
    unclassified: Money


class NatureData(BaseModel):
    series: list[NatureMonth]
    # NOTE: fixtures.py's `nature()` never populates a `reference` key —
    # the mock's build() only calls nature() with no reference argument
    # and the dict it returns has no such field. contratos.html's "Forma
    # do dado" for block 4 does list `reference: {...}|None` at the top
    # level though, so we keep it here as the documented contract but
    # default it to None since the fixtures never exercise it.
    reference: NatureReference | None = None


# ---------------------------------------------------------------------------
# 5 - Projection
# ---------------------------------------------------------------------------


class ProjectionComponents(BaseModel):
    income_expected: Money
    recurring: Money
    installments: Money
    variable_estimate: Money


class ProjectionPoint(BaseModel):
    month: str
    kind: Literal["actual", "projected"]
    balance: Money
    low: Money | None
    high: Money | None
    committed: Money
    components: ProjectionComponents


class ProjectionAssumption(BaseModel):
    label: str
    value: str
    source: str


class ProjectionData(BaseModel):
    points: list[ProjectionPoint]
    assumptions: list[ProjectionAssumption]


# ---------------------------------------------------------------------------
# 6 - Categories table
# ---------------------------------------------------------------------------


class CategoryRow(BaseModel):
    category_id: str | None
    label: str
    color: str
    group_label: str
    nature: Literal["fixed", "variable", "discretionary"] | None
    amount: Money
    reference: Money | None
    reference_method: str | None
    delta: Money | None
    delta_pct: float | None
    tx_count: int
    status: Literal["over", "under", "ok", "no_ref"]


class CategoriesData(BaseModel):
    rows: list[CategoryRow]
    available: bool
    blocked_reason: str | None


class CategoriesByReference(BaseModel):
    """fixtures.py nests both reference modes under one response:
    dict(categories=dict(budget=categories("budget"), historical=categories("historical")))
    NOTE: this is a mock-build-time convenience (both modes computed once
    for the whole scenario dump), not necessarily the real endpoint shape —
    contratos.html's per-endpoint contract for `/api/insights/categories`
    takes a `reference` query param and returns a single CategoriesData,
    matching the brief. We model both: CategoriesData for the real
    per-request endpoint, and this wrapper only if a combined shape is
    ever needed. The real endpoint should return InsightsEnvelope[CategoriesData].
    """

    budget: CategoriesData
    historical: CategoriesData


# ---------------------------------------------------------------------------
# 7 - Hygiene
# ---------------------------------------------------------------------------


class HygieneCoverageStat(BaseModel):
    done: int
    total: int

    # The screen never computes. `fixtures.py` emits only {done, total} and
    # leaves the division to the client, which is the defect the review filed
    # as G8 — the mock was never corrected, so the fixtures carry it forward.
    # Deriving `pct` here rather than storing it keeps it from drifting away
    # from the counts it summarizes.
    @computed_field  # type: ignore[prop-decorator]
    @property
    def pct(self) -> float:
        if self.total == 0:
            return 0.0
        return round(self.done / self.total * 100, 1)


class HygieneCoverage(BaseModel):
    categorized: HygieneCoverageStat
    nature_set: HygieneCoverageStat
    transfers_paired: HygieneCoverageStat


class HygieneOffender(BaseModel):
    kind: Literal["month", "account", "payee"]
    label: str
    missing: int
    amount: Money


class HygieneData(BaseModel):
    coverage: HygieneCoverage
    last_review_at: datetime | None
    # NOTE: fixtures.py's hygiene() never emits `days_since_review` at all —
    # the dict returned is exactly
    # dict(coverage=..., last_review_at=..., worst_offenders=...).
    # The brief explicitly asks for a `days_since_review: int` field
    # ("client must never divide or subtract dates"), and contratos.html's
    # "Forma do dado" for block 7 lists it too
    # (`days_since_review: int|null`). We keep the field since two of the
    # three sources require it and it's clearly intended, but note that
    # the fixtures themselves never populate it — likely an oversight in
    # the mock rather than a deliberate omission.
    days_since_review: int | None = None
    worst_offenders: list[HygieneOffender]


# ---------------------------------------------------------------------------
# 8 - Goals
# ---------------------------------------------------------------------------


class GoalRow(BaseModel):
    goal_id: str
    label: str
    current: Money
    target: Money
    progress: float
    target_date: date | None
    observed_contribution: Money | None
    estimated_completion: date | None
    required_contribution: Money | None
    status: Literal["on_track", "behind", "stalled", "unknown"]
    blocked_reason: str | None


# `goals` in fixtures.py is a bare list.
GoalsData = list[GoalRow]


# ---------------------------------------------------------------------------
# 9 - Breakeven table + purchase decision calculator
# ---------------------------------------------------------------------------


class YieldRange(BaseModel):
    p25: str
    p75: str


class YieldWindowDays(BaseModel):
    min: int
    max: int


class YieldBasis(BaseModel):
    monthly_gross: str
    method: Literal["observed", "assumed"]
    sample_size: int
    window_days: YieldWindowDays
    range: YieldRange
    as_of: date


class BreakevenRow(BaseModel):
    n: int
    ir_rate: float
    net_monthly: str
    breakeven_discount: float
    gain_per_1000: Money


class BreakevenTableData(BaseModel):
    yield_basis: YieldBasis
    rows: list[BreakevenRow]


class PurchaseDecisionRequest(BaseModel):
    price: Money
    cash_discount_pct: float
    installments: int
    yield_override: float | None = None


class PurchaseDecisionVerdict(BaseModel):
    choice: Literal["installments", "cash"]
    net_gain: Money
    breakeven_discount: float
    headline: str


class PurchaseDecisionScheduleRow(BaseModel):
    month: int
    payment: Money
    balance_invested: Money
    yield_net: Money
    cumulative_gain: Money


class FlexibilityMonthlyCommitmentDelta(BaseModel):
    month: str
    # NOTE: contratos.html's schema sketch for
    # monthly_commitment_delta is `[{month, ...}]` — the ellipsis means the
    # remaining fields aren't specified anywhere (not in the brief, not in
    # fixtures.py, which doesn't build purchase-decision responses at all;
    # that endpoint is a POST computed live, outside the fixtures dump).
    # Modeling the one concrete field we have plus a delta amount as the
    # most plausible reading; this needs confirmation once the backend
    # calculator is actually specified.
    delta: Money


class FlexibilityImpact(BaseModel):
    utilization_before: float
    utilization_after: float
    monthly_commitment_delta: list[FlexibilityMonthlyCommitmentDelta]
    cash_gap_month: str | None


class PurchaseDecisionData(BaseModel):
    verdict: PurchaseDecisionVerdict
    schedule: list[PurchaseDecisionScheduleRow]
    flexibility_impact: FlexibilityImpact
